### PR TITLE
Automate doxygen builds

### LINKED ISSUE #5: HIDAPI web site
Someone should work on a web site presenting the hidapi project and subprojects (like Python wrapper).
goals:
* make the new hidapi project more visible so that it is indexed by search engine
* document the history from https://github.com/signal11/hidapi

One technical solution is to use github pages https://pages.github.com/ and host the web site at https://hidapi.libusb.info/ (the name does not exist yet but I can create it)

### PR DESCRIPTION
**Pull request by 2bndy5**
resolves #5 

- updated the doxyfile using `doxygen -u`
- include hidtest/test.c as a doc'd example
- use a brief `doxygen/main_page.md` file as the generated site's main page.
- strip absolute paths from doxygen output (which now displays as relative paths to repo root)
- added CI to build the docs and save as a CI artifact (useful for visual inspection without checking out a PR branch). This new CI will also rewrite the contents of the `gh-pages` branch for deployment (only for push events to the master branch).
- added some missing docstrings and excluded undocumented "helper macros" to satisfy warnings output in doxygen's build log.

## Admin prep

The required preparation needed from admins is
1. make sure the [repo's setting](https://github.com/libusb/hidapi/settings/pages) are set to use the `gh-pages` branch root folder as source of deployments. 
[image](https://docs.github.com/articles/configuring-a-publishing-source-for-github-pages/)
   I realize that there is already a branch named `github_pages`, but I think that branch was meant for a different purpose. The branch published to github pages typically only has the site's HTML resources (not any programming or documentation sources) -- especially when using Doxygen as the static site generator.
2. make sure the [custom domain](https://docs.github.com/articles/using-a-custom-domain-with-github-pages/) is configured correctly. I mention this one because https://libusb.github.io/hidapi forwards to https://libusb.info/hidapi
3. add the deployed site's link to the repo's "About" settings: 
image
Alternatively (or additionally), a link could be added to the README in the ["What does the API look like?" section](https://github.com/libusb/hidapi#what-does-the-api-look-like).

### CODE REVIEW

**Comment by Youw** [comment]
I've made a job on master branch, to test CI in this PR.
As for other changes/requests - I'll get to it tiny bit later. Too much preoccupied at the moment.

**Comment by 2bndy5** [comment]
> I realize that there is already a branch named github_pages, but I think that branch was meant for a different purpose. The branch published to github pages typically only has the site's HTML resources (not any programming or documentation sources) -- especially when using Doxygen as the static site generator.

The branch used to publish the docs' site can be configured ([see here](https://github.com/peaceiris/actions-gh-pages#%EF%B8%8F-set-another-github-pages-branch-publish_branch)) Default is a branch named `gh-pages`.

File: doxygen/main_page.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+This site is dedicated to hosting an [API reference for the HIDAPI library](#API).
+
+For general information, see the [source repository](https://github.com/libusb/hidapi#readme).
+
+There are also build instructions hosted on github:
+
+- [Building from source](https://github.com/2bndy5/hidapi/blob/master/BUILD.md)
+- [Using CMake](https://github.com/2bndy5/hidapi/blob/master/BUILD.cmake.md)
+- [Using Autotools (deprecated)](https://github.com/2bndy5/hidapi/blob/master/BUILD.autotools.md)
```
Comment (Youw) [inline doxygen/main_page.md:11]: updaye links

File: doxygen/Doxyfile
Comment (Youw) [inline doxygen/Doxyfile:1]: it always blows my mind how many changes this file gets when it gets updated by doxygen

File: hidapi/hidapi.h [hunk 2]
```diff
@@ -103,10 +105,11 @@
...
 extern "C" {
 #endif
+		/** A structure to hold the version numbers. */
 		struct hid_api_version {
-			int major;
-			int minor;
-			int patch;
+			int major; ///< major version number
+			int minor; ///< minor version number
+			int patch; ///< patch version number
```
Comment (Youw) [inline hidapi/hidapi.h:112]: as per agreement - we use only `/** c-style comments */` - for consisteny

File: hidapi/hidapi.h
Comment (Youw) [inline hidapi/hidapi.h:1]: generally less changes than I thought there would be 
thanks

File: .github/workflows/docs.yaml [hunk 3]
```diff
@@ -13,3 +13,42 @@ jobs:
...
+        uses: actions/download-artifact@v3
+        with:
+          name: HIDAPI_doxygen_docs
+          path: docs
+
+      - name: upload to github pages
+        uses: peaceiris/actions-gh-pages@v3
+        with:
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+          publish_dir: ./docs
```
Comment (Youw) [inline .github/workflows/docs.yaml:54]: I don't think the history of documentation changes has any value in the branch, given that all doc sources are in the repo anyway
```suggestion
          publish_dir: ./docs
          force_orphan: true
```

**Comment by Youw** [comment]
> already a branch named `github_pages`, but I think that branch was meant for a different purpose

originally it was created for the same purpose as a playground
we just didn't make it right to the end
at this moment I believe it is time to get rid of it

File: .github/workflows/docs.yaml [hunk 4]
```diff
@@ -13,3 +13,42 @@ jobs:
...
+        uses: actions/upload-artifact@v3
+        with:
+          name: HIDAPI_doxygen_docs
+          path: ${{ github.workspace }}/doxygen/html
+
+  deploy-docs:
+    runs-on: ubuntu-latest
+    needs: [build]
+    if: github.ref_type == 'branch' && github.ref_name == 'master'
+    steps:
```
Comment (Youw) [inline .github/workflows/docs.yaml:43]: lets make sure to deploy only the latest
```suggestion
    concurrency:
      group: "github-pages-deploy"
      cancel-in-progress: true
    steps:
```

**Comment by Youw** [comment]
image

File: doxygen/main_page.md (hunk 1 above)
Comment (2bndy5) [reply doxygen/main_page.md:11]: Oops! 😆

File: hidapi/hidapi.h (hunk 2 above)
Comment (2bndy5) [reply hidapi/hidapi.h:112]: Oh, I couldn't find any code style guideline, but I didn't look very hard...

Comment (2bndy5) [reply hidapi/hidapi.h]: done. The `<` is required as part of doxygen trailing docstring syntax.

File: hidapi/hidapi.h
Comment (2bndy5) [reply hidapi/hidapi.h:1]: Yep, that was the intention. I see another PR that makes changes to the hidapi.h file... FYI, I didn't make the CI fail if doxygen emits a warning (typically when a docstring isn't provided for something).

File: doxygen/Doxyfile
Comment (2bndy5) [reply doxygen/Doxyfile:1]: TBH, I prefer only having settings in the Doxyfile that don't use the default value. Although, I didn't go through that laborious change here.

File: doxygen/Doxyfile [hunk 5]
```diff
@@ -865,25 +985,27 @@ EXCLUDE_PATTERNS       =
...
-# AClass::ANamespace, ANamespace::*Test
+# ANamespace::AClass, ANamespace::*Test
 #
 # Note that the wildcards are matched against the file with absolute path, so to
 # exclude all test directories use the pattern */test/*
 
-EXCLUDE_SYMBOLS        =
+EXCLUDE_SYMBOLS        = HID_API_AS_STR_IMPL \
+                         HID_API_AS_STR \
+                         HID_API_TO_VERSION_STR
```
Comment (2bndy5) [inline doxygen/Doxyfile:995]: This is the important part that allows eliding specific members from the generated docs. ✏️ Here, I used it to exclude the "helper macros" that don't have a docstring.

File: hidapi/hidapi.h (hunk 2 above)
Comment (Youw) [reply hidapi/hidapi.h]: > couldn't find any code style guideline

unfortunately those aren't written anywhere
just the fact of consistency

File: hidapi/hidapi.h
Comment (Youw) [reply hidapi/hidapi.h:1]: makes sense at this point, thanks

File: doxygen/main_page.md [hunk 6]
```diff
@@ -0,0 +1,13 @@
+# HIDAPI Doxygen output
+
+This site is dedicated to hosting an [API reference for the HIDAPI library](#API).
```
Comment (Youw) [inline doxygen/main_page.md:3]: I'm not sure where `#API` link points to

Comment (Youw) [reply doxygen/main_page.md:3]: ah, I see - that's a Doxygen link

**Pull request merged**